tests/Unit/Middleware/TrimMiddlewareTest.php

```diff
@@ -0,0 +1,96 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Engelsystem\Test\Unit\Middleware;
+
+use Engelsystem\Http\Request;
+use Engelsystem\Middleware\TrimMiddleware;
+use Engelsystem\Test\Unit\TestCase;
+use Generator;
+use PHPUnit\Framework\MockObject\MockObject;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Server\RequestHandlerInterface;
+
+class TrimMiddlewareTest extends TestCase
+{
+    private TrimMiddleware $subject;
+
+    /**
+     * @param MockObject<RequestHandlerInterface>
+     */
+    private MockObject $handler;
+
+    public function setUp(): void
+    {
+        $this->subject = new TrimMiddleware();
+        $this->handler = $this->getMockForAbstractClass(RequestHandlerInterface::class);
+    }
+
+    /**
+     * @return Generator<string, array>
+     */
+    public function provideTrimTestData(): Generator
+    {
+        yield 'GET request' => ['GET', [], []];
+
+        foreach (['POST', 'PUT'] as $method) {
+            yield $method . ' request with empty data' => [$method, [], []];
+            yield $method . ' request with mixed data' => [
+                $method,
+                [
+                    'fieldA' => 23,
+                    'fieldB' => ' bla ',
+                    'password' => ' pass1 ',
+                    'password2' => ' pass2 ',
+                    'fieldC' => ['sub' => ' bla2 '],
+                    'fieldD' => null,
+                ],
+                [
+                    'fieldA' => 23,
+                    'fieldB' => 'bla',
+                    // password fields should keep their surrounding spaces
+                    'password' => ' pass1 ',
+                    'password2' => ' pass2 ',
+                    'fieldC' => ['sub' => 'bla2'],
+                    'fieldD' => null,
+                ],
+            ];
+        }
+    }
+
+    /**
+     * @covers \Engelsystem\Middleware\TrimMiddleware
+     * @dataProvider provideTrimTestData
+     */
+    public function testTrim(string $method, mixed $body, mixed $expectedBody): void
+    {
+        $request = (new Request())->withMethod($method)->withParsedBody($body);
+        $this->handler->expects(self::once())->method('handle')->with(
+            self::callback(function (ServerRequestInterface $request) use ($expectedBody): bool {
+                self::assertSame($expectedBody, $request->getParsedBody());
+                return true;
+            })
+        );
+        $this->subject->process($request, $this->handler);
+    }
+
+    /**
+     * Special test case to cover null value parsed body.
+     *
+     * @covers \Engelsystem\Middleware\TrimMiddleware
+     */
+    public function testTrimPostNull(): void
+    {
+        $request = $this->getMockForAbstractClass(ServerRequestInterface::class);
```

Comment:
Imho creating "real" requests / responses normally makes tests more readable / future proof as they don't need to be updated/that many changes when refactoring the code itself

Reply:
Doesn't work here, because the underlying implementation doesn't allow `null` as parsed body. This is allowed [according to the spec](https://www.php-fig.org/psr/psr-7/).

> 1) Engelsystem\Test\Unit\Middleware\TrimInputTest::testTrimPostNull
TypeError: Symfony\Component\HttpFoundation\InputBag::replace(): Argument #1 ($inputs) must be of type array, null given, called in /home/weeman/git/chaos/engelsystem/src/Http/Request.php on line 392